File: ui/shared/context.go
// Copyright (c) 2026 lazybeeper by Ronen Druker.

package shared

// PanelFocus identifies which panel currently has keyboard focus.
type PanelFocus int

const (
	// FocusAccounts is the accounts list panel.
	FocusAccounts PanelFocus = iota
	// FocusChats is the chats list panel.
	FocusChats
	// FocusMessages is the message viewport panel.
	FocusMessages
	// FocusInput is the message input textarea.
	FocusInput
	// FocusPopup is an active popup/overlay.
	FocusPopup
)

// panelCount is the number of navigable panels (excluding popup).
const panelCount = 4

// String returns the human-readable name of the panel focus.
func (f PanelFocus) String() string {
	switch f {
	case FocusAccounts:
		return "Accounts"
	case FocusChats:
		return "Chats"
	case FocusMessages:
		return "Messages"
	case FocusInput:
		return "Input"
	case FocusPopup:
		return "Popup"
	default:
		return "Unknown"
	}
}

// NextPanel returns the next panel in the Tab cycle order.
func (f PanelFocus) NextPanel() PanelFocus {
	if f == FocusPopup {
		return FocusAccounts
	}

	return PanelFocus((int(f) + 1) % panelCount)
}

// PrevPanel returns the previous panel in the Shift+Tab cycle order.
func (f PanelFocus) PrevPanel() PanelFocus {
	if f == FocusPopup {
		return FocusInput
	}

	return PanelFocus((int(f) + panelCount - 1) % panelCount)
}

// RightPanel returns the panel to the right. From sidebar panels,
// moves to messages. From messages/input, stays put.
func (f PanelFocus) RightPanel() PanelFocus {
	switch f {
	case FocusAccounts, FocusChats:
		return FocusMessages
	case FocusMessages, FocusInput, FocusPopup:
		return f
	}

	return f
}

// LeftPanel returns the panel to the left. From main area, moves
// to chats. From sidebar, stays put.
func (f PanelFocus) LeftPanel() PanelFocus {
	switch f {
	case FocusMessages, FocusInput:
		return FocusChats
	case FocusAccounts, FocusChats, FocusPopup:
		return f
	}

	return f
}

File: src/ui/viewmodel/context.ts
// Copyright (c) 2026 lazybeeper by Ronen Druker.

/** Identifies which panel currently has keyboard focus. */
export enum PanelFocus {
  /** Accounts list panel. */
  Accounts = 0,
  /** Chats list panel. */
  Chats = 1,
  /** Message viewport panel. */
  Messages = 2,
  /** Message input textarea. */
  Input = 3,
  /** Active popup/overlay. */
  Popup = 4,
}

/** Number of navigable panels (excluding popup). */
const PANEL_COUNT = 4;

/**
 * Returns the human-readable name of the panel focus.
 * @param focus - The current panel focus value.
 * @returns The display name of the panel.
 */
export function panelFocusName(focus: PanelFocus): string {
  switch (focus) {
    case PanelFocus.Accounts:
      return "Accounts";
    case PanelFocus.Chats:
      return "Chats";
    case PanelFocus.Messages:
      return "Messages";
    case PanelFocus.Input:
      return "Input";
    case PanelFocus.Popup:
      return "Popup";
  }
}

/**
 * Returns the next panel in the Tab cycle order.
 * @param focus - The current panel focus.
 * @returns The next panel in the cycle.
 */
export function nextPanel(focus: PanelFocus): PanelFocus {
  if (focus === PanelFocus.Popup) {
    return PanelFocus.Accounts;
  }

  return ((focus + 1) % PANEL_COUNT) as PanelFocus;
}

/**
 * Returns the previous panel in the Shift+Tab cycle order.
 * @param focus - The current panel focus.
 * @returns The previous panel in the cycle.
 */
export function prevPanel(focus: PanelFocus): PanelFocus {
  if (focus === PanelFocus.Popup) {
    return PanelFocus.Input;
  }

  return ((focus + PANEL_COUNT - 1) % PANEL_COUNT) as PanelFocus;
}

/**
 * Returns the panel to the right. From sidebar, moves to messages.
 * @param focus - The current panel focus.
 * @returns The panel to the right, or the same panel if no right neighbor.
 */
export function rightPanel(focus: PanelFocus): PanelFocus {
  switch (focus) {
    case PanelFocus.Accounts:
    case PanelFocus.Chats:
      return PanelFocus.Messages;
    default:
      return focus;
  }
}

/**
 * Returns the panel to the left. From main area, moves to chats.
 * @param focus - The current panel focus.
 * @returns The panel to the left, or the same panel if no left neighbor.
 */
export function leftPanel(focus: PanelFocus): PanelFocus {
  switch (focus) {
    case PanelFocus.Messages:
    case PanelFocus.Input:
      return PanelFocus.Chats;
    default:
      return focus;
  }
}
